docs/design/features/data-stream.md

```diff
@@ -0,0 +1,106 @@
+
+**NOTE** Registration of types should be done prior to any recording of instances. It is assumed that all sharable types are known statically.
+
+After type registration is performed, streams can be aquired by components in the target process and typed instances inserted into the stream. Adding an instance to a stream is considered thread safe. The typing of an instance should be done via a numeric ID.
+
+### Reader usage
+
+The reader should first define a series of type names that it is able to interpret and consume. These type names should match the names defined by the target. These names could be used to map types in the target process with their numeric ID and version. The reader is not expected to have any hardcoded numeric type IDs as these are subject to change between target versions.
+
+The reader is expected to be resiliant in recieving an unknown version of a type and gracefully interpret it. Two examples of graceful interpretation are describing it as `"Unknown"` or printing its memory address in the target process.
+
+After acquiring the target process's `data_stream_context_t` address, it should be validated and endianness computed.
+
+The first time the data stream is read, the target processes types are enumerated in reserve chronological order (Last In/First Out) and this data may be used throughout the lifetime of the target process. During the enumeration of types the following can be done:
+
+* Creation of a fast mapping from name to numeric ID.
+
+* Creation of look-up map to type details (e.g., field offsets).
+
+* Validation of supported type versions.
+
+After type enumeration is complete, instance streams can be enumerated and interpreted. The size contained within the type description allows the reader to read in the entire type and then use field offsets to poke into that memory. The reading in of the entire data type helps with efficiency and acts as a versioning resiliance mechanism. Adding new fields to a type, without changing the version, need not represent a breaking change.
+
+### Design FAQs
+
+---
+
+**Q1** Why are streams allowed to grow?
+
+**A1** Consider the case where a data structure in the target process has a specific use case but the reader has either stricter or looser requirements. An example would be a thread pool used in the target process. This structure would ideally only be concerned with current threads in the target process, exited threads having been removed. However, the reader process likely has a need for knowing when a thread instance has exited to update its own internal state. A possible solution is to fully query the thread pool data structure each time. However, if instead entries for created and deleted threads are recorded in a stream, the reader only needs to know the delta as opposed to querying the thread pool each time. The logic follows for any data structure that contains objects with transient lifetimes.
```

Comment:
I think we should flip the question around a bit. Lets assume by definition that streams are data structures that support growth. Do we expect a global list of types and a global list of instances are scenarios that need growth? Both NativeAOT and CoreCLR have similar lists, neither of them currently support growth, and it has never been an issue. From my POV trying to use streams for that purpose seems to be introducing extra complexity in the data contract where a flat array would suffice.

Where growth appears to enter the picture is the assumption that we should use the same list both to encode global singletons ([examples](https://github.com/dotnet/runtime/blob/main/src/coreclr/inc/dacvars.h)) and also dynamically created objects of interest (such as all Threads). I don't believe we are getting any benefit by combining them though.

> However, if instead entries for created and deleted threads are recorded in a stream, the reader only needs to know the delta as opposed to querying the thread pool each time.

I don't think we should be maintaining a list of every deleted thread because the potential size of that list is unbounded. I'd advocate that memory usage in this feature must remain close to what it currently is regardless of app behavior. If an app wants to create and destroy threads continuously (threadpool with variable load may do just that) then that app needs to keep working.

It seems like there is an underlying assumption in the design that streams are a generally good data structure that we will be used repeatedly for all sorts of different data and I don't agree with that premise.

Reply:
> It seems like there is an underlying assumption in the design that streams are a generally good data structure that we will be used repeatedly for all sorts of different data and I don't agree with that premise.

The streams idea may be a useful concept for certain contracts (in particular contracts oriented around live-debugging). I agree with Noah that the streams idea does not seem to be applicable for the scenarios that we are going after first (#99298).

Reply:
> I don't think we should be maintaining a list of every deleted thread because the potential size of that list is unbounded.

We are not intending to do this. It was an example (perhaps poor) of the potential flexibility. I'd agree we currently don't have a specific initial use case for that flexibility.

Reply:
> does not seem to be applicable for the scenarios that we are going after first (https://github.com/dotnet/runtime/issues/99298).

I think this is a slight confusion and we keep focusing on this, which I don't fully understand. Note that the `data_stream` library itself was written as a contract that is not concerned with the contents. It can be dynamic or static - one initializes with a series of sizes (dynamic) the other could accept a large block of memory (static, but not implemented). The decision here to abstract that away in a library was precisely to avoid conflating dynamic vs static concerns. If we had avoided the data stream abstraction, it likely would force us down assumptions about the static nature, this design avoids that mistake with the abstraction. @elinor-fung mentions this above as well, it is an implementation detail that has costs we may not want to accept right now, however in the current form we have more degrees of freedom with the abstraction than without.

Reply:
I do not understand how the streams fit into the minidump scenarios. 

If we go with the streams idea, what are going to be the streams saved into the minidump?

Reply:
> [Elinor] I'd agree we currently don't have a specific initial use case for that flexibility.

Any objections to waiting until we do have such a use case before introducing it?


> [Aaron] I think this is a slight confusion and we keep focusing on this, which I don't fully understand... The decision here to abstract that away in a library was precisely to avoid conflating dynamic vs static concerns.

The reason I focus on it is because abstracting it behind a library API doesn't resolve all the concerns:
- One of the major outputs of this feature is a specification for a data format. That format spec is intended to allow anyone to create their own parser for this data. Parsing code in one library doesn't preclude developers from needing to work directly with the underlying data.
- The choice of data structures directly impacts performance characteristics such as application size on disk, application size in memory, dump size on disk, and data access time.
- A growing stream vs. a static stream have different rules for when the data needs to be re-queried. This means callers will still care which one they are dealing with regardless of the code that does a point-in-time enumeration.